### ISSUE TITLE
Is an update of the Standard File Protocol Specification planned any time soon?

### ISSUE THREAD

**Issue by Zegnat** [standardfile/standardfile.github.io#12]
I was looking into Standard Notes and really wanted to know what the server was handling and if I could make use of Standard File myself. Thought I would use my own preferred programming language by implementing [the spec](https://standardfile.org/), but it seems the spec is woefully out of date with what Standard Notes is doing and expecting.

I am looking at Standard Notes as a reference implementation of a client here, it does the following things that are undocumented just within the registration and login flows:

1. Sends an undocumented `version` parameter on `POST /auth`.
2. Expects an undocumented `user` parameter back from `POST /auth`.
3. Expects an undocumented `version` parameter back from `GET /auth/params`.
4. Expects an undocumented `user` parameter back from `POST /auth/sign_in`.

Now I haven’t even gotten to synchronisation yet, but #9 suggests there will be parameters missing there as well.

This is especially bad as [the 2017 security audit](https://standardnotes.org/blog/7/announcing-our-2017-security-audit-results) only covered features as described on `standardfile.org`. If swathes of required back-and-forth between client and server were excluded from the audit, the audit is incomplete. I realise a `version` parameter more or less isn’t going to make your server less secure, but I also do not know what other parameters might be missing as I haven’t reviewed Standard Notes nor a current server implementation. And the auditors haven’t done so either.

(I was also surprised to read that there are apparently 2 versions of Standard File v0.0.2, one from before the audit and one from after? What value has a `version` parameter then?)

**Comment by mobitar** [comment]
Actually, yes, an update is planned. Not a version update. Just better documentation. Not just for Standard File but the entire Standard Notes ecosystem. This is what I'm working on as we speak.

It's worth noting that the Standard File server only deals with encrypted data, and has very little moving parts. It was designed to be very dumb, with the intention of protecting us from making unwelcome changes. 

The `version` parameter was actually added in tandem with the audit, which is why it's not reflected.

Really what it boils down to is missing documentation of the `user` parameter, which [only returns](https://github.com/standardfile/ruby-server/blob/master/app/models/user.rb#L6) the user's email and uuid, which I overlooked. Thanks for the heads up.

`conflict_of` flag was also removed as part of the improvements laid out by the audit. I just updated https://github.com/standardfile/standardfile.github.io/issues/9#issuecomment-352266491 with details.

So I wouldn't call if _woefully_ out of date, but does not need some touching up.

**Comment by Zegnat** [comment]
Let me start by apologising for the “woefully” remark, that was completely unnecessarily throwing shade and not part of the problem at all. I need to get better at not doing that and appreciate you calling me out on it.

The fact I was calling it out of date was partially because #9 had been open for half a year and I never checked if the concerns there (another undocumented parameter) were still applicable. That is addressed now 🎉

This issue was mostly a problem of perception on my part. Which is great to hear! I hope the touch ups will show up soon as I am still interested in developing my own back-end to synchronise against. I prefer implementing a spec over porting code from e.g. Ruby.

> It's worth noting that the Standard File server only deals with encrypted data, and has very little moving parts. It was designed to be very dumb, with the intention of protecting us from making unwelcome changes.

Like I said, I didn’t really think any of these extra parameters would have an effect on the security findings of the audit. But it might be worth including non-normative notes in the spec that call out changes based on the audit / made after the audit. This would address all possible confusion about having a public v0.0.2, getting this v0.0.2 audited, and then publishing a revised v0.0.2.

E.g. when including `user` have a note saying “Note: this parameter was not part of the original v0.0.2 audit and is not important to security sensitive parts of the specification.” (Or any other wording.)

**Comment by mobitar** [comment]
Haha no worries, trust me, I do the same sometimes. 

Also, good points about how to best structure documentation updates. 

> possible confusion about having a public v0.0.2, getting this v0.0.2 audited, and then publishing a revised v0.0.2.

Though just to be clear: Version 0.0.2 was the result of the audit, and not publicly available before then. So basically, we had version 001, which was unaudited and used in production. Then design work began on 002. When the spec was complete, I immediately began working with the auditors to make sure it was safe before beginning to use 002 in production. 

In any case, what I'm working on now will be a central repository for all things documentation. Hope to have some updates within the week. Thanks again for the feedback.

**Comment by Zegnat** [comment]
> Though just to be clear: Version 0.0.2 was the result of the audit, and not publicly available before then.

That’s good then!

Honestly this concern came to be because of [the report (PDF)](https://s3.amazonaws.com/standard-notes/security/Report-SN-Audit.pdf). The report specifically states it is reviewing “v0.0.2, retrieved from https://standardfile.org/ on May 17, 2017.” Which makes it sound like the spec was already published. The August updates then says they “retrieved” a new version of v0.0.2 in June, which made it sound like this was a second version of v0.0.2. Not having looked at the production development of the Ruby server or the Standard Notes application this gave me the wrong idea about the spec.

Not sure if this really requires addressing in the spec, as only people reading the audit might get confused, but at least you’ll have an idea where my assumptions came from.

**Comment by mobitar** [comment]
Yeah, totally. I've been building horizontally for the last year and so some verticals are incomplete, like documentation. The 002 spec was published for the purpose of the audit but unused. I may have taken for granted just how much the public can follow along on all this. But thus why my current mission is very thorough documentation.
